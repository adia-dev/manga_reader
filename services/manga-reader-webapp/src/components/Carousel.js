import React, { useEffect, useState } from "react";
import {
  BiCaretDown,
  BiChevronDown,
  BiChevronRight,
  BiChevronUp,
} from "react-icons/bi";
import {
  BsArrowDown,
  BsArrowDownShort,
  BsArrowLeftShort,
  BsArrowRightShort,
  BsArrowUpShort,
  BsFillMouseFill,
} from "react-icons/bs";
import MangaData from "../data/mangas.json";

const Carousel = () => {
  const [current, setCurrent] = useState(0);
  const [currentPage, setCurrentPage] = useState(0);
  const [startMouseY, setStartMouseY] = useState(0);
  const [startMouseX, setStartMouseX] = useState(0);
  const [mouseY, setMouseY] = useState(0);
  const [mouseX, setMouseX] = useState(0);
  const [isDragging, setIsDragging] = useState(false);
  const [mangas, setMangas] = useState(MangaData);
  const [canPlayTrailer, setCanPlayTrailer] = useState(false);
  const [canPlayTrailerTimeout, setCanPlayTrailerTimeout] = useState(null);

  // load manga videos with require

  useEffect(() => {
    const interval = setInterval(() => {
      if (isDragging) return;
      setCurrent((current) => {
        if (current === mangas.length - 1) {
          return 0;
        } else {
          return current + 1;
        }
      });
    }, 5000);

    const handleKeyDown = (e) => {
      if (e.key === "ArrowUp") {
        setCurrent(Math.max(current - 1, 0));
      } else if (e.key === "ArrowDown") {
        setCurrent(Math.min(current + 1, mangas.length - 1));
      }

      if (e.key === "ArrowLeft") {
        setCurrentPage(Math.max(currentPage - 1, 0));
      } else if (e.key === "ArrowRight") {
        setCurrentPage(
          Math.min(currentPage + 1, mangas[current].page_count - 1)
        );
      }
    };
    window.addEventListener("keydown", handleKeyDown);

    return () => {
      clearInterval(interval);
      window.removeEventListener("keydown", handleKeyDown);
    };
  }, [mangas, current, isDragging, currentPage]);

  const SCROLL_TRESHOLD = 150;

  const MouseDeltaY = () => mouseY - startMouseY;
  const MouseDeltaX = () => mouseX - startMouseX;

  const AbsMouseDeltaY = () => Math.abs(MouseDeltaY());
  const AbsMouseDeltaX = () => Math.abs(MouseDeltaX());

  function minmax(val, min, max) {
    return Math.min(Math.max(val, min), max);
  }

  // vertical carousel
  const OnMouseDown = (e) => {
    setIsDragging(true);
    setStartMouseY(e.clientY);
    setStartMouseX(e.clientX);
    setMouseY(e.clientY);
    setMouseX(e.clientX);

    if (canPlayTrailerTimeout) {
      clearTimeout(canPlayTrailerTimeout);
      setCanPlayTrailerTimeout(null);
      setCanPlayTrailer(false);
    }

    setCanPlayTrailerTimeout(
      setTimeout(() => {
        setCanPlayTrailer(true);
      }, 1500)
    );
  };

  const OnMouseMove = (e) => {
    if (!isDragging) return;

    setMouseY(e.clientY);
    setMouseX(e.clientX);
  };

  const OnMouseUpOrLeave = (e) => {
    if (!isDragging) return;

    setIsDragging(false);
    setMouseY(e.clientY);
    setMouseX(e.clientX);
    setStartMouseY(e.clientY);
    setStartMouseX(e.clientX);

    if (canPlayTrailerTimeout) {
      clearTimeout(canPlayTrailerTimeout);
    }
    setCanPlayTrailerTimeout(null);
    setCanPlayTrailer(false);

    if (mouseY - startMouseY > SCROLL_TRESHOLD) {
      setCurrent(Math.max(current - 1, 0));
      return;
    } else if (mouseY - startMouseY < -SCROLL_TRESHOLD) {
      setCurrent(Math.min(current + 1, mangas.length - 1));
      return;
    }

    if (mouseX - startMouseX > SCROLL_TRESHOLD) {
      setCurrentPage(Math.max(currentPage - 1, 0));
      return;
    } else if (mouseX - startMouseX < -SCROLL_TRESHOLD) {
      setCurrentPage(Math.min(currentPage + 1, mangas[current].page_count - 1));
      return;
    }
  };

  return (
    <div
      className="w-screen h-screen overflow-hidden bg-gray-900 text-white"
      id="carousel"
    >
      <div className="w-screen h-screen overflow-hidden flex flex-col p-10 items-center justify-center absolute top-0 left-0 z-10 pointer-events-none">
        <div
          className="absolute w-full top-0 left-0 flex justify-between px-6 py-14 transition duration-1000 delay-300 ease-in-out"
          style={{
            transform: `scale(${isDragging ? 0.9 : 1.5}, 1)`,
          }}
        >
          <div className="flex flex-col items-center space-y-2">
            <BsFillMouseFill className="text-2xl" />
            <div className="flex flex-col items-center text-black">
              <div className="w-6 h-6 rounded-md border bg-gray-200 grid place-items-center m-1">
                <BsArrowUpShort className="" />
              </div>
              <div className="flex items-center">
                <div className="w-6 h-6 rounded-md border bg-gray-200 grid place-items-center m-1">
                  <BsArrowLeftShort className="" />
                </div>
                <div className="w-6 h-6 rounded-md border bg-gray-200 grid place-items-center m-1">
                  <BsArrowDownShort className="" />
                </div>
                <div className="w-6 h-6 rounded-md border bg-gray-200 grid place-items-center m-1">
                  <BsArrowRightShort className="" />
                </div>
              </div>
            </div>
            <p className="text-xs text-gray-300">Mouse and Keyboard support</p>
          </div>
          <p className="text-xs text-gray-300">
            {currentPage + 1}/{mangas[current].page_count}
          </p>
        </div>
        {current > 0 && (
          <div
            className="flex flex-col items-center absolute top-0 transition delay-200 ease-out duration-1000"
            style={{
              transform: `translateY(${
                isDragging ? 100 : -200
              }%) scale(${minmax(MouseDeltaY() / 100, 0.8, 1)})`,
              color: MouseDeltaY() > SCROLL_TRESHOLD ? "green" : "white",
            }}
          >
            <div className="flex flex-col items-center animate-bounce">
              <BiChevronUp className="text-5xl" />
              <p>{mangas[current - 1].title}</p>
            </div>
          </div>
        )}
        {current < mangas.length - 1 && (
          <div
            className="absolute bottom-0 transition delay-200 ease-out duration-1000"
            style={{
              transform: `translateY(${isDragging ? -100 : 200}%) scale(${
                MouseDeltaY() ? 0.8 : minmax(MouseDeltaY() / 100, 0.8, 1)
              })`,
              color: MouseDeltaY() < -SCROLL_TRESHOLD ? "green" : "white",
            }}
          >
            <div className="flex flex-col items-center animate-bounce">
              <p>{mangas[current + 1].title}</p>
              <BiChevronDown className="text-5xl" />
            </div>
          </div>
        )}
        {mangas[current].page_count > 1 && (
          <div
            className="flex items-center justify-center absolute right-0 mb-10 mr-1 transition delay-200 ease-out duration-1000"
            style={{
              transform: `translateX(${isDragging ? -100 : 200}%)`,
              color: MouseDeltaY() < -SCROLL_TRESHOLD ? "green" : "white",
            }}
          >
            <p className="text-xs mr-2">{mangas[current].page_count} pages</p>
            <BiChevronRight className="text-2xl" />
          </div>
        )}
      </div>
      {mangas.map((manga, index) => {
        return (
          <div
            className="flex flex-col overflow-hidden items-center justify-center h-full relative transition duration-500 delay-100 ease-in-out transform cursor-pointer group"
            onMouseDown={OnMouseDown}
            onMouseMove={OnMouseMove}
            onMouseUp={OnMouseUpOrLeave}
            onMouseLeave={OnMouseUpOrLeave}
            style={{
              transform: `${
                isDragging
                  ? `translate(
                    ${currentPage * -100}%, 
                    math(${current * -100}%) + ${MouseDeltaY()}px)`
                  : `translate(
                    ${currentPage * -100}%,
                    ${current * -100}%)`
              }`,
              filter: `blur(${minmax(
                AbsMouseDeltaY() / 50,
                0,
                10
              )}px) brightness(${1 - minmax(AbsMouseDeltaY() / 1000, 0, 0.5)})`,
            }}
            key={manga.id}
            data-testid="manga"
          >
            <div className="w-full h-full overflow-hidden absolute top-0 left-0 z-0  rounded-3xl scale-105 cursor-pointer group-active:scale-90 transition duration-500 ease-in-out delay-100">
              <div className="absolute flex items-center justify-center w-full h-full ">
                <img
                  src={manga.image}
                  alt="manga"
                  className="w-full h-full brightness-75 group-active:brightness-50 transition-all object-cover "
                  style={{
                    opacity: index === current ? 1 : 0,
                  }}
                  draggable="false"
                />
                <div className="flex flex-col items-center justify-center h-full z-10 not-selectable absolute">
                  <h1 className="text-5xl font-bold">{manga.title}</h1>
                  <p className="text-gray-100">Chapter: {manga.chapter}</p>
                </div>
              </div>
              {current === index && isDragging && canPlayTrailer && (
                <video
                  className="flex items-center justify-center w-full h-full object-cover absolute top-0 left-0 z-10"
                  autoPlay
                  loop
                  src={require(`../assets/videos/trailers/${manga.trailerID}.mp4`)}
                ></video>
              )}
            </div>
            <img
              src={manga.image}
              alt="manga"
              className="w-full h-full brightness-95 object-cover absolute overflow-hidden top-0 left-0 -z-10 blur-3xl scale-150 transition duration-500 ease-in-out delay-100"
              style={{
                opacity: index === current ? 1 : 0,
              }}
              draggable="false"
            />
          </div>
        );
      })}
    </div>
  );
};

export default Carousel;
