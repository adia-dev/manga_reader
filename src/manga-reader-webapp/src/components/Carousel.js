import React, { useEffect, useState } from "react";
import { BiChevronDown, BiChevronRight, BiChevronUp } from "react-icons/bi";

const Carousel = () => {
  const [current, setCurrent] = useState(0);
  const [startMouseY, setStartMouseY] = useState(0);
  const [mouseY, setMouseY] = useState(0);
  const [isDragging, setIsDragging] = useState(false);
  const [mangas, setMangas] = useState([
    {
      id: 1,
      title: "Jujutsu Kaisen",
      image: "https://images5.alphacoders.com/114/1141137.jpg",
      page_count: 18,
      chapter: 1,
    },
    {
      id: 2,
      title: "One Piece",
      image:
        "https://preview.redd.it/cujgnglafov81.png?width=640&crop=smart&auto=webp&s=bff4be0d66177d9fd3a8d75a8a02a21f00cfb950",
      page_count: 17,
      chapter: 1017,
    },
    {
      id: 3,
      title: "Bleach",
      image:
        "https://c4.wallpaperflare.com/wallpaper/109/45/304/bleach-thousand-year-blood-war-arc-tite-kubo-gotei-13-4k-hd-wallpaper-preview.jpg",
      page_count: 16,
      chapter: 684,
    },
    {
      id: 4,
      title: "Hell's Paradise: Jigokuraku",
      image: "https://animeland.fr/wp-content/uploads/2021/01/hellp.jpg",
      page_count: 15,
      chapter: 56,
    },
  ]);

  useEffect(() => {
    const interval = setInterval(() => {
      if (isDragging) return;
      setCurrent((current) => {
        if (current === mangas.length - 1) {
          return 0;
        } else {
          return current + 1;
        }
      });
    }, 5000);

    const handleKeyDown = (e) => {
      if (e.key === "ArrowUp") {
        setCurrent(Math.max(current - 1, 0));
      } else if (e.key === "ArrowDown") {
        setCurrent(Math.min(current + 1, mangas.length - 1));
      }
    };
    window.addEventListener("keydown", handleKeyDown);

    return () => {
      clearInterval(interval);
      window.removeEventListener("keydown", handleKeyDown);
    };
  }, [mangas, current, isDragging]);

  const SCROLL_TRESHOLD = 150;

  const MouseDelta = () => mouseY - startMouseY;
  const AbsMouseDelta = () => Math.abs(MouseDelta());

  function minmax(val, min, max) {
    return Math.min(Math.max(val, min), max);
  }

  // vertical carousel
  const OnMouseDown = (e) => {
    setIsDragging(true);
    setStartMouseY(e.clientY);
    setMouseY(e.clientY);
  };

  const OnMouseMove = (e) => {
    if (!isDragging) return;

    setMouseY(e.clientY);
  };

  const OnMouseUpOrLeave = (e) => {
    if (!isDragging) return;

    setIsDragging(false);
    setMouseY(e.clientY);
    setStartMouseY(e.clientY);

    if (mouseY - startMouseY > SCROLL_TRESHOLD) {
      setCurrent(Math.max(current - 1, 0));
    } else if (mouseY - startMouseY < -SCROLL_TRESHOLD) {
      setCurrent(Math.min(current + 1, mangas.length - 1));
    }
  };

  return (
    <div
      className="w-screen h-screen overflow-hidden bg-gray-900 text-white"
      id="carousel"
    >
      <div className="w-screen h-screen overflow-hidden flex flex-col p-10 items-center justify-center absolute top-0 left-0 z-10 pointer-events-none">
        {current > 0 && (
          <div
            className="flex flex-col items-center absolute top-0 transition delay-200 ease-out duration-1000"
            style={{
              transform: `translateY(${
                isDragging ? 100 : -200
              }%) scale(${minmax(MouseDelta() / 100, 0.8, 1)})`,
              color: MouseDelta() > SCROLL_TRESHOLD ? "green" : "white",
            }}
          >
            <div className="flex flex-col items-center animate-bounce">
              <BiChevronUp className="text-5xl" />
              <p>{mangas[current - 1].title}</p>
            </div>
          </div>
        )}
        {current < mangas.length - 1 && (
          <div
            className="absolute bottom-0 transition delay-200 ease-out duration-1000"
            style={{
              transform: `translateY(${isDragging ? -100 : 200}%) scale(${
                MouseDelta() ? 0.8 : minmax(MouseDelta() / 100, 0.8, 1)
              })`,
              color: MouseDelta() < -SCROLL_TRESHOLD ? "green" : "white",
            }}
          >
            <div className="flex flex-col items-center animate-bounce">
              <p>{mangas[current + 1].title}</p>
              <BiChevronDown className="text-5xl" />
            </div>
          </div>
        )}
        {mangas[current].page_count > 1 && (
          <div
            className="flex items-center justify-center absolute right-0 mb-10 mr-1 transition delay-200 ease-out duration-1000"
            style={{
              transform: `translateX(${isDragging ? 0 : 200}%)`,
              color: MouseDelta() < -SCROLL_TRESHOLD ? "green" : "white",
            }}
          >
            <p className="text-xs mr-2">{mangas[current].page_count} pages</p>
            <BiChevronRight className="text-2xl" />
          </div>
        )}
      </div>
      {mangas.map((manga, index) => {
        return (
          <div
            className="flex flex-col items-center justify-center h-full relative transition duration-500 delay-100 ease-in-out transform cursor-pointer group"
            onMouseDown={OnMouseDown}
            onMouseMove={OnMouseMove}
            onMouseUp={OnMouseUpOrLeave}
            onMouseLeave={OnMouseUpOrLeave}
            style={{
              transform: `${
                isDragging
                  ? `translateY(math(${current * -100}%) + ${MouseDelta()}px)`
                  : `translateY(${current * -100}%)`
              }`,
              filter: `blur(${minmax(
                AbsMouseDelta() / 50,
                0,
                10
              )}px) brightness(${1 - minmax(AbsMouseDelta() / 1000, 0, 0.5)})`,
            }}
            key={manga.id}
          >
            <img
              src={manga.image}
              alt="manga"
              className="w-full h-full brightness-95 object-cover absolute top-0 left-0 z-0 group-active:brightness-50 rounded-3xl scale-105 group-active:scale-90 transition duration-500 ease-in-out delay-100"
              style={{
                opacity: index === current ? 1 : 0,
              }}
              draggable="false"
            />
            {/* 
                an other image on the background that is scaled up and blurred 
            */}

            <img
              src={manga.image}
              alt="manga"
              className="w-full h-full brightness-95 object-cover absolute overflow-hidden top-0 left-0 -z-10 blur-3xl scale-150 transition duration-500 ease-in-out delay-100"
              style={{
                opacity: index === current ? 1 : 0,
              }}
              draggable="false"
            />

            <div className="flex flex-col items-center justify-center h-full z-10 not-selectable">
              <h1 className="text-5xl font-bold">{manga.title}</h1>
              <p className="text-gray-100">Chapter: {manga.chapter}</p>
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default Carousel;
