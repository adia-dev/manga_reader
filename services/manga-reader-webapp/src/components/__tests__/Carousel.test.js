import Carousel from "../Carousel";
import { render, screen } from "@testing-library/react";
const mangas = require("../../data/mangas.json");

describe("Carousel component related tests", () => {
  beforeEach(() => {});

  it("Should render the carousel component", () => {
    render(<Carousel />);
    const carousel = screen.getByTestId("carousel");
    expect(carousel).toBeInTheDocument();
  });

  it("Should render the carousel component with the correct number of manga", async () => {
    render(<Carousel />);
    const mangaList = await screen.findAllByTestId("carousel-item");
    expect(mangaList.length).toBe(mangas.length);
  });

  it("Should render the titles of the manga in the page", async () => {
    const { asFragment } = render(<Carousel />);
    mangas.forEach((manga) => {
      expect(asFragment()).toHaveTextContent(manga.title);
    });
  });

  it("Should render the user instructions correctly", () => {
    render(<Carousel />);
    const instructions = screen.getByText("Mouse and Keyboard support");
    expect(instructions).toBeInTheDocument();
  });

  it("Should render the UI when the user clicks and holds the left mouse button", () => {
    render(<Carousel />);
    const carousel = screen.getAllByTestId("carousel-item")[0];
    const event = new MouseEvent("mousedown", {
      bubbles: true,
      cancelable: true,
      button: 0,
    });
    carousel.dispatchEvent(event);
    expect(carousel).toHaveClass("grabbing");
  });
});
