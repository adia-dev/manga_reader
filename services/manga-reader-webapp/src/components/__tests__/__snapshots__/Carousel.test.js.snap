// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component related tests Should render the carousel component 1`] = `
<div
  className="w-screen h-screen overflow-hidden bg-gray-900 text-white"
  id="carousel"
>
  <div
    className="w-screen h-screen overflow-hidden flex flex-col p-10 items-center justify-center absolute top-0 left-0 z-10 pointer-events-none"
  >
    <div
      className="absolute w-full top-0 left-0 flex justify-between px-6 py-14 transition duration-1000 delay-300 ease-in-out"
      style={
        Object {
          "transform": "scale(1.5, 1)",
        }
      }
    >
      <div
        className="flex flex-col items-center space-y-2"
      >
        <svg
          className="text-2xl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 5a5 5 0 0 1 10 0v6a5 5 0 0 1-10 0V5zm5.5-1.5a.5.5 0 0 0-1 0v2a.5.5 0 0 0 1 0v-2z"
          />
        </svg>
        <div
          className="flex flex-col items-center text-black"
        >
          <div
            className="w-6 h-6 rounded-md border bg-gray-200 grid place-items-center m-1"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 12a.5.5 0 0 0 .5-.5V5.707l2.146 2.147a.5.5 0 0 0 .708-.708l-3-3a.5.5 0 0 0-.708 0l-3 3a.5.5 0 1 0 .708.708L7.5 5.707V11.5a.5.5 0 0 0 .5.5z"
                fillRule="evenodd"
              />
            </svg>
          </div>
          <div
            className="flex items-center"
          >
            <div
              className="w-6 h-6 rounded-md border bg-gray-200 grid place-items-center m-1"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8a.5.5 0 0 1-.5.5H5.707l2.147 2.146a.5.5 0 0 1-.708.708l-3-3a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L5.707 7.5H11.5a.5.5 0 0 1 .5.5z"
                  fillRule="evenodd"
                />
              </svg>
            </div>
            <div
              className="w-6 h-6 rounded-md border bg-gray-200 grid place-items-center m-1"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 4a.5.5 0 0 1 .5.5v5.793l2.146-2.147a.5.5 0 0 1 .708.708l-3 3a.5.5 0 0 1-.708 0l-3-3a.5.5 0 1 1 .708-.708L7.5 10.293V4.5A.5.5 0 0 1 8 4z"
                  fillRule="evenodd"
                />
              </svg>
            </div>
            <div
              className="w-6 h-6 rounded-md border bg-gray-200 grid place-items-center m-1"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 8a.5.5 0 0 1 .5-.5h5.793L8.146 5.354a.5.5 0 1 1 .708-.708l3 3a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708-.708L10.293 8.5H4.5A.5.5 0 0 1 4 8z"
                  fillRule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
        <p
          className="text-xs text-gray-300"
        >
          Mouse and Keyboard support
        </p>
      </div>
      <p
        className="text-xs text-gray-300"
      >
        1
        /
        18
      </p>
    </div>
    <div
      className="absolute bottom-0 transition delay-200 ease-out duration-1000"
      style={
        Object {
          "color": "white",
          "transform": "translateY(200%) scale(0.8)",
        }
      }
    >
      <div
        className="flex flex-col items-center animate-bounce"
      >
        <p>
          One Piece
        </p>
        <svg
          className="text-5xl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.293 9.293 12 13.586 7.707 9.293l-1.414 1.414L12 16.414l5.707-5.707z"
          />
        </svg>
      </div>
    </div>
    <div
      className="flex items-center justify-center absolute right-0 mb-10 mr-1 transition delay-200 ease-out duration-1000"
      style={
        Object {
          "color": "white",
          "transform": "translateX(200%)",
        }
      }
    >
      <p
        className="text-xs mr-2"
      >
        18
         pages
      </p>
      <svg
        className="text-2xl"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.707 17.707 16.414 12l-5.707-5.707-1.414 1.414L13.586 12l-4.293 4.293z"
        />
      </svg>
    </div>
  </div>
  <div
    className="flex flex-col overflow-hidden items-center justify-center h-full relative transition duration-500 delay-100 ease-in-out transform cursor-pointer group"
    data-testid="manga"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "filter": "blur(0px) brightness(1)",
        "transform": "translate(
                    0%,
                    0%)",
      }
    }
  >
    <div
      className="w-full h-full overflow-hidden absolute top-0 left-0 z-0  rounded-3xl scale-105 cursor-pointer group-active:scale-90 transition duration-500 ease-in-out delay-100"
    >
      <div
        className="absolute flex items-center justify-center w-full h-full "
      >
        <img
          alt="manga"
          className="w-full h-full brightness-75 group-active:brightness-50 transition-all object-cover "
          draggable="false"
          src="https://images5.alphacoders.com/114/1141137.jpg"
          style={
            Object {
              "opacity": 1,
            }
          }
        />
        <div
          className="flex flex-col items-center justify-center h-full z-10 not-selectable absolute"
        >
          <h1
            className="text-5xl font-bold"
          >
            Jujutsu Kaisen
          </h1>
          <p
            className="text-gray-100"
          >
            Chapter: 
            1
          </p>
        </div>
      </div>
    </div>
    <img
      alt="manga"
      className="w-full h-full brightness-95 object-cover absolute overflow-hidden top-0 left-0 -z-10 blur-3xl scale-150 transition duration-500 ease-in-out delay-100"
      draggable="false"
      src="https://images5.alphacoders.com/114/1141137.jpg"
      style={
        Object {
          "opacity": 1,
        }
      }
    />
  </div>
  <div
    className="flex flex-col overflow-hidden items-center justify-center h-full relative transition duration-500 delay-100 ease-in-out transform cursor-pointer group"
    data-testid="manga"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "filter": "blur(0px) brightness(1)",
        "transform": "translate(
                    0%,
                    0%)",
      }
    }
  >
    <div
      className="w-full h-full overflow-hidden absolute top-0 left-0 z-0  rounded-3xl scale-105 cursor-pointer group-active:scale-90 transition duration-500 ease-in-out delay-100"
    >
      <div
        className="absolute flex items-center justify-center w-full h-full "
      >
        <img
          alt="manga"
          className="w-full h-full brightness-75 group-active:brightness-50 transition-all object-cover "
          draggable="false"
          src="https://preview.redd.it/cujgnglafov81.png?width=640&crop=smart&auto=webp&s=bff4be0d66177d9fd3a8d75a8a02a21f00cfb950"
          style={
            Object {
              "opacity": 0,
            }
          }
        />
        <div
          className="flex flex-col items-center justify-center h-full z-10 not-selectable absolute"
        >
          <h1
            className="text-5xl font-bold"
          >
            One Piece
          </h1>
          <p
            className="text-gray-100"
          >
            Chapter: 
            1017
          </p>
        </div>
      </div>
    </div>
    <img
      alt="manga"
      className="w-full h-full brightness-95 object-cover absolute overflow-hidden top-0 left-0 -z-10 blur-3xl scale-150 transition duration-500 ease-in-out delay-100"
      draggable="false"
      src="https://preview.redd.it/cujgnglafov81.png?width=640&crop=smart&auto=webp&s=bff4be0d66177d9fd3a8d75a8a02a21f00cfb950"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
  </div>
  <div
    className="flex flex-col overflow-hidden items-center justify-center h-full relative transition duration-500 delay-100 ease-in-out transform cursor-pointer group"
    data-testid="manga"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "filter": "blur(0px) brightness(1)",
        "transform": "translate(
                    0%,
                    0%)",
      }
    }
  >
    <div
      className="w-full h-full overflow-hidden absolute top-0 left-0 z-0  rounded-3xl scale-105 cursor-pointer group-active:scale-90 transition duration-500 ease-in-out delay-100"
    >
      <div
        className="absolute flex items-center justify-center w-full h-full "
      >
        <img
          alt="manga"
          className="w-full h-full brightness-75 group-active:brightness-50 transition-all object-cover "
          draggable="false"
          src="https://c4.wallpaperflare.com/wallpaper/109/45/304/bleach-thousand-year-blood-war-arc-tite-kubo-gotei-13-4k-hd-wallpaper-preview.jpg"
          style={
            Object {
              "opacity": 0,
            }
          }
        />
        <div
          className="flex flex-col items-center justify-center h-full z-10 not-selectable absolute"
        >
          <h1
            className="text-5xl font-bold"
          >
            Bleach
          </h1>
          <p
            className="text-gray-100"
          >
            Chapter: 
            684
          </p>
        </div>
      </div>
    </div>
    <img
      alt="manga"
      className="w-full h-full brightness-95 object-cover absolute overflow-hidden top-0 left-0 -z-10 blur-3xl scale-150 transition duration-500 ease-in-out delay-100"
      draggable="false"
      src="https://c4.wallpaperflare.com/wallpaper/109/45/304/bleach-thousand-year-blood-war-arc-tite-kubo-gotei-13-4k-hd-wallpaper-preview.jpg"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
  </div>
  <div
    className="flex flex-col overflow-hidden items-center justify-center h-full relative transition duration-500 delay-100 ease-in-out transform cursor-pointer group"
    data-testid="manga"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "filter": "blur(0px) brightness(1)",
        "transform": "translate(
                    0%,
                    0%)",
      }
    }
  >
    <div
      className="w-full h-full overflow-hidden absolute top-0 left-0 z-0  rounded-3xl scale-105 cursor-pointer group-active:scale-90 transition duration-500 ease-in-out delay-100"
    >
      <div
        className="absolute flex items-center justify-center w-full h-full "
      >
        <img
          alt="manga"
          className="w-full h-full brightness-75 group-active:brightness-50 transition-all object-cover "
          draggable="false"
          src="https://animeland.fr/wp-content/uploads/2021/01/hellp.jpg"
          style={
            Object {
              "opacity": 0,
            }
          }
        />
        <div
          className="flex flex-col items-center justify-center h-full z-10 not-selectable absolute"
        >
          <h1
            className="text-5xl font-bold"
          >
            Hell's Paradise: Jigokuraku
          </h1>
          <p
            className="text-gray-100"
          >
            Chapter: 
            56
          </p>
        </div>
      </div>
    </div>
    <img
      alt="manga"
      className="w-full h-full brightness-95 object-cover absolute overflow-hidden top-0 left-0 -z-10 blur-3xl scale-150 transition duration-500 ease-in-out delay-100"
      draggable="false"
      src="https://animeland.fr/wp-content/uploads/2021/01/hellp.jpg"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
  </div>
</div>
`;
